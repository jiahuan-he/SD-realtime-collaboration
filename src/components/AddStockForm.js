import React from 'react';

const button = {
    backgroundColor: "white",
    border: "1px solid",
    color: "black",
    padding: "5px 12px",
    cursor: "pointer",
}

const inputInvalid = {
    border: "2px solid #ed6663"
}

export default class AddStockForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            stockName:"",
            stockValue: 0,
        };
    }

    handleChangeName = (event) => {
        this.setState({ stockName: event.target.value.trim()});
    }

    handleChangeValue = (event) => {
        this.setState({ stockValue: event.target.value});
    }

    isValidStockID = () => {    
        if(!this.state.stockName) return false
        return this.props.stocks.filter((stock => stock.id === this.state.stockName)).length === 0 
        && this.props.flows.filter((flow => flow.id === this.state.stockName)).length === 0
    }

    isValidStockValue =() => {
        return !isNaN(this.state.stockValue)
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <label>
                    Stock Name
                <input 
                    style={this.isValidStockID()?null:inputInvalid} 
                    type="text" 
                    value={this.state.stockName} 
                    onChange={this.handleChangeName} 
                />
                </label>
                <label>
                    Init Value
                <input 
                    style={this.isValidStockValue()?null:inputInvalid} 
                    type="text" 
                    value={this.state.stockValue} onChange={this.handleChangeValue} />
                </label>
                    <input type="button" value="Add Stock" style={button}
                        onClick={() => {
                            if(this.state.stockName === "" || !this.isValidStockID() || !this.isValidStockValue()) return
                            this.props.addStock(this.state.stockName, this.state.stockValue)
                            this.setState({
                                stockName:"",
                                stockValue: 0
                            })
                        }}
                    />
            </form>
        )
    }
}