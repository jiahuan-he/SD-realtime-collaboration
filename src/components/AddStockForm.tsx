import React, {FormEvent, CSSProperties } from 'react';
import { I_Stock, I_Flow, I_Parameter } from '../model';

interface State{
    stockName: string
    stockValue: number
}

interface Props{
    stocks: Array<I_Stock>,
    flows: Array<I_Flow>,
    parameters: Array<I_Parameter>,
    addStock: (id:string, value:number)=>void,
    button: CSSProperties,
    inputInvalid: CSSProperties,
}

export default class AddStockForm extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            stockName:"",
            stockValue: 0,
        };
    }

    handleChangeName = (event: FormEvent<HTMLInputElement>) => {
        this.setState({ stockName: event.currentTarget.value.trim()});
    }

    handleChangeValue = (event: FormEvent<HTMLInputElement>) => {
        this.setState({ stockValue: Number(event.currentTarget.value)});
    }

    isValidStockID = () => {
        if(!this.state.stockName) return false
        return this.props.stocks.filter((stock => stock.id === this.state.stockName)).length === 0 
        && this.props.flows.filter((flow => flow.id === this.state.stockName)).length === 0
        && this.props.parameters.filter((parameter => parameter.name === this.state.stockName)).length === 0
    }

    isValidStockValue =() => {
        return !isNaN(this.state.stockValue)
    }

    render() {
        return (
            <form>
                <label>
                    Stock Name
                <input 
                    style={this.isValidStockID()?undefined:this.props.inputInvalid} 
                    type="text" 
                    value={this.state.stockName} 
                    onChange={this.handleChangeName} 
                />
                </label>
                <label>
                    Init Value
                <input 
                    style={this.isValidStockValue()?undefined:this.props.inputInvalid} 
                    type="text" 
                    value={this.state.stockValue} onChange={this.handleChangeValue} />
                </label>
                    <input type="button" value="Add Stock" style={this.props.button}
                        onClick={() => {
                            if(this.state.stockName === "" || !this.isValidStockID() || !this.isValidStockValue()) return
                            this.props.addStock(this.state.stockName, this.state.stockValue)
                            this.setState({
                                stockName:"",
                                stockValue: 0
                            })
                        }}
                    />
            </form>
        )
    }
}