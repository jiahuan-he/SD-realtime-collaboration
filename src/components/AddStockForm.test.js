import React from 'react';
import { shallow } from 'enzyme';
import AddStockForm from './AddStockForm';
describe('AddStockForm', () => {
    it('should render correctly in "debug" mode', () => {
        const props = {
            "stockIDs": ["stock1", "stock2"],
        };
        const form = shallow(<AddStockForm
            stockIDs={props.stockIDs}
            debug
        ></AddStockForm>);
        expect(form).toMatchSnapshot();
    });
});

describe('AddStockForm', () => {
    it('should render correctly in "debug" mode', () => {
        const props = {
            "stockIDs": [],
        };
        const form = shallow(<AddStockForm
            stockIDs={props.stockIDs}
            debug
        ></AddStockForm>);
        expect(form).toMatchSnapshot();
    });
});