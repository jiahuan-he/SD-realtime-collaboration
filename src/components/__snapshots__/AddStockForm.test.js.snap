// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddStockForm should render correctly in "debug" mode 1`] = `
<form>
  <label>
    Stock Name
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value=""
    />
  </label>
  <label>
    Value
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value={0}
    />
  </label>
  <input
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "border": "1px solid",
        "color": "black",
        "cursor": "pointer",
        "padding": "5px 12px",
      }
    }
    type="button"
    value="Add Stock"
  />
</form>
`;

exports[`AddStockForm should render correctly in "debug" mode 2`] = `
<form>
  <label>
    Stock Name
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value=""
    />
  </label>
  <label>
    Value
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value={0}
    />
  </label>
  <input
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "border": "1px solid",
        "color": "black",
        "cursor": "pointer",
        "padding": "5px 12px",
      }
    }
    type="button"
    value="Add Stock"
  />
</form>
`;
